### ISSUE TITLE
404 errors in Autoscaler logs when Job purge or Garbage collection occurs

### ISSUE THREAD

**Issue by dhung-hashicorp** [hashicorp/nomad-autoscaler#604]
**Nomad Autoscaler version**
0.3.7+ent (also happens on earlier versions)

**Operating system and Environment details**
Alpine Linux 3.11.13

**Issue**
When a Nomad job with autoscaler scaling policy is cleaned up via `nomad job -purge` or `nomad system gc`, there appears to be a race condition between the GC and internal autoscaler cache of policy. As a result, these 404s errors are showing up in the autoscaler logs:

[ERROR] policy_manager.policy_handler: encountered an error monitoring policy: policy_id=0c66157a-0e7b-fdc9-de1d-5fb429d14439 error="failed to get policy: Unexpected response code: 404 (policy not found)"
[ERROR] internal_plugin.app-sizing-nomad: failed to call Nomad API: job_id=example namespace=default error="Unexpected response code: 404 (job not found)"
[DEBUG] internal_plugin.nomad-target: failed to read job scale status, retrying: Job=example error="Unexpected response code: 404 (job not found)"

**Reproduction steps**
Run the job spec listed below. Then perform `nomad job stop -purge 

**Job spec:**
{
  "Stop": false,
  "Region": "global",
  "Namespace": "default",
  "ID": "example",
  "ParentID": "",
  "Name": "example",
  "Type": "service",
  "Priority": 50,
  "AllAtOnce": false,
  "Datacenters": [
    "dc1"
  ],
  "Constraints": null,
  "Affinities": null,
  "Spreads": null,
  "TaskGroups": [
    {
      "Name": "cache-lb",
      "Count": 1,
      "Update": {
        "Stagger": 30000000000,
        "MaxParallel": 1,
        "HealthCheck": "checks",
        "MinHealthyTime": 10000000000,
        "HealthyDeadline": 300000000000,
        "ProgressDeadline": 600000000000,
        "AutoRevert": false,
        "AutoPromote": false,
        "Canary": 0
      },
      "Migrate": {
        "MaxParallel": 1,
        "HealthCheck": "checks",
        "MinHealthyTime": 10000000000,
        "HealthyDeadline": 300000000000
      },
      "Constraints": [
        {
          "LTarget": "${attr.os.signals}",
          "RTarget": "SIGHUP",
          "Operand": "set_contains"
        },
        {
          "LTarget": "${attr.consul.version}",
          "RTarget": ">= 1.7.0",
          "Operand": "semver"
        }
      ],
      "Scaling": null,
      "RestartPolicy": {
        "Attempts": 2,
        "Interval": 1800000000000,
        "Delay": 15000000000,
        "Mode": "fail"
      },
      "Tasks": [
        {
          "Name": "nginx",
          "Driver": "docker",
          "User": "",
          "Config": {
            "volumes": [
              "local/nginx.conf:/etc/nginx/nginx.conf",
              "local/nginx:/etc/nginx/conf.d"
            ],
            "image": "nginx",
            "ports": [
              "lb"
            ]
          },
          "Env": null,
          "Services": null,
          "Vault": null,
          "Templates": [
            {
              "SourcePath": "",
              "DestPath": "local/nginx.conf",
              "EmbeddedTmpl": "user  nginx;\nworker_processes  1;\n\nerror_log  /var/log/nginx/error.log warn;\npid        /var/run/nginx.pid;\n\n\nevents {\n    worker_connections  1024;\n}\n\ninclude /etc/nginx/conf.d/*.conf;\n",
              "ChangeMode": "restart",
              "ChangeSignal": "",
              "Splay": 5000000000,
              "Perms": "0644",
              "Uid": -1,
              "Gid": -1,
              "LeftDelim": "{{",
              "RightDelim": "}}",
              "Envvars": false,
              "VaultGrace": 0,
              "Wait": null
            },
            {
              "SourcePath": "",
              "DestPath": "local/nginx/nginx.conf",
              "EmbeddedTmpl": "stream {\n  server {\n    listen {{ env \"NOMAD_PORT_lb\" }};\n    proxy_pass backend;\n  }\n\n  upstream backend {\n  {{ range service \"redis\" }}\n    server {{ .Address }}:{{ .Port }};\n  {{ else }}server 127.0.0.1:65535; # force a 502\n  {{ end }}\n  }\n}\n",
              "ChangeMode": "signal",
              "ChangeSignal": "SIGHUP",
              "Splay": 5000000000,
              "Perms": "0644",
              "Uid": -1,
              "Gid": -1,
              "LeftDelim": "{{",
              "RightDelim": "}}",
              "Envvars": false,
              "VaultGrace": 0,
              "Wait": null
            }
          ],
          "Constraints": null,
          "Affinities": null,
          "Resources": {
            "CPU": 57,
            "Cores": 0,
            "MemoryMB": 10,
            "MemoryMaxMB": 0,
            "DiskMB": 0,
            "IOPS": 0,
            "Networks": null,
            "Devices": null
          },
          "RestartPolicy": {
            "Attempts": 2,
            "Interval": 1800000000000,
            "Delay": 15000000000,
            "Mode": "fail"
          },
          "DispatchPayload": null,
          "Lifecycle": null,
          "Meta": null,
          "KillTimeout": 5000000000,
          "LogConfig": {
            "MaxFiles": 10,
            "MaxFileSizeMB": 10
          },
          "Artifacts": null,
          "Leader": false,
          "ShutdownDelay": 0,
          "VolumeMounts": null,
          "ScalingPolicies": [
            {
              "ID": "0ad8e3f2-1554-608f-0128-b6b834e935aa",
              "Type": "vertical_cpu",
              "Target": {
                "Namespace": "default",
                "Job": "example",
                "Group": "cache-lb",
                "Task": "nginx"
              },
              "Policy": {
                "check": [
                  {
                    "95pct": [
                      {
                        "source": "prometheus",
                        "strategy": [
                          {
                            "app-sizing-percentile": [
                              {
                                "percentile": "95"
                              }
                            ]
                          }
                        ]
                      }
                    ]
                  }
                ],
                "cooldown": "1m",
                "evaluation_interval": "1m"
              },
              "Min": 0,
              "Max": -1,
              "Enabled": true,
              "CreateIndex": 0,
              "ModifyIndex": 0
            },
            {
              "ID": "3e211765-4432-a648-53e5-e1d89f16fba9",
              "Type": "vertical_mem",
              "Target": {
                "Task": "nginx",
                "Namespace": "default",
                "Job": "example",
                "Group": "cache-lb"
              },
              "Policy": {
                "cooldown": "1m",
                "evaluation_interval": "1m",
                "check": [
                  {
                    "max": [
                      {
                        "source": "prometheus",
                        "strategy": [
                          {
                            "app-sizing-max": [
                              {}
                            ]
                          }
                        ]
                      }
                    ]
                  }
                ]
              },
              "Min": 0,
              "Max": -1,
              "Enabled": true,
              "CreateIndex": 0,
              "ModifyIndex": 0
            }
          ],
          "KillSignal": "",
          "Kind": "",
          "CSIPluginConfig": null
        }
      ],
      "EphemeralDisk": {
        "Sticky": false,
        "SizeMB": 300,
        "Migrate": false
      },
      "Meta": null,
      "ReschedulePolicy": {
        "Attempts": 0,
        "Interval": 0,
        "Delay": 30000000000,
        "DelayFunction": "exponential",
        "MaxDelay": 3600000000000,
        "Unlimited": true
      },
      "Affinities": null,
      "Spreads": null,
      "Networks": [
        {
          "Mode": "",
          "Device": "",
          "CIDR": "",
          "IP": "",
          "Hostname": "",
          "MBits": 0,
          "DNS": null,
          "ReservedPorts": null,
          "DynamicPorts": [
            {
              "Label": "lb",
              "Value": 0,
              "To": 0,
              "HostNetwork": "default"
            }
          ]
        }
      ],
      "Consul": {
        "Namespace": ""
      },
      "Services": [
        {
          "Name": "redis-lb",
          "TaskName": "",
          "PortLabel": "lb",
          "AddressMode": "host",
          "Address": "",
          "EnableTagOverride": false,
          "Tags": null,
          "CanaryTags": null,
          "Checks": [
            {
              "Name": "service: \"redis-lb\" check",
              "Type": "tcp",
              "Command": "",
              "Args": null,
              "Path": "",
              "Protocol": "",
              "PortLabel": "lb",
              "Expose": false,
              "AddressMode": "",
              "Interval": 10000000000,
              "Timeout": 2000000000,
              "InitialStatus": "",
              "TLSSkipVerify": false,
              "Method": "",
              "Header": null,
              "CheckRestart": null,
              "GRPCService": "",
              "GRPCUseTLS": false,
              "TaskName": "",
              "SuccessBeforePassing": 0,
              "FailuresBeforeCritical": 0,
              "Body": "",
              "OnUpdate": "require_healthy"
            }
          ],
          "Connect": null,
          "Meta": null,
          "CanaryMeta": null,
          "Namespace": "default",
          "OnUpdate": "require_healthy",
          "Provider": "consul"
        }
      ],
      "Volumes": null,
      "ShutdownDelay": null,
      "StopAfterClientDisconnect": null,
      "MaxClientDisconnect": null
    },
    {
      "Name": "cache",
      "Count": 1,
      "Update": {
        "Stagger": 30000000000,
        "MaxParallel": 1,
        "HealthCheck": "checks",
        "MinHealthyTime": 10000000000,
        "HealthyDeadline": 300000000000,
        "ProgressDeadline": 600000000000,
        "AutoRevert": false,
        "AutoPromote": false,
        "Canary": 0
      },
      "Migrate": {
        "MaxParallel": 1,
        "HealthCheck": "checks",
        "MinHealthyTime": 10000000000,
        "HealthyDeadline": 300000000000
      },
      "Constraints": [
        {
          "LTarget": "${attr.consul.version}",
          "RTarget": ">= 1.7.0",
          "Operand": "semver"
        }
      ],
      "Scaling": {
        "ID": "3d406e1e-e590-936b-a68a-a2d563f029ba",
        "Type": "horizontal",
        "Target": {
          "Group": "cache",
          "Namespace": "default",
          "Job": "example"
        },
        "Policy": {
          "check": [
            {
              "CPU": [
                {
                  "source": "datadog",
                  "query": "nomad.client.allocs.cpu.total_ticks{task:redis}",
                  "strategy": [
                    {
                      "target-value": [
                        {
                          "target": 50
                        }
                      ]
                    }
                  ]
                }
              ]
            }
          ],
          "cooldown": "1m",
          "evaluation_interval": "30s"
        },
        "Min": 1,
        "Max": 9,
        "Enabled": true,
        "CreateIndex": 0,
        "ModifyIndex": 0
      },
      "RestartPolicy": {
        "Attempts": 2,
        "Interval": 1800000000000,
        "Delay": 15000000000,
        "Mode": "fail"
      },
      "Tasks": [
        {
          "Name": "redis",
          "Driver": "docker",
          "User": "",
          "Config": {
            "image": "redis:6.0",
            "ports": [
              "db"
            ]
          },
          "Env": null,
          "Services": [
            {
              "Name": "redis",
              "TaskName": "",
              "PortLabel": "db",
              "AddressMode": "host",
              "Address": "",
              "EnableTagOverride": false,
              "Tags": null,
              "CanaryTags": null,
              "Checks": [
                {
                  "Name": "service: \"redis\" check",
                  "Type": "tcp",
                  "Command": "",
                  "Args": null,
                  "Path": "",
                  "Protocol": "",
                  "PortLabel": "db",
                  "Expose": false,
                  "AddressMode": "",
                  "Interval": 10000000000,
                  "Timeout": 2000000000,
                  "InitialStatus": "",
                  "TLSSkipVerify": false,
                  "Method": "",
                  "Header": null,
                  "CheckRestart": null,
                  "GRPCService": "",
                  "GRPCUseTLS": false,
                  "TaskName": "",
                  "SuccessBeforePassing": 0,
                  "FailuresBeforeCritical": 0,
                  "Body": "",
                  "OnUpdate": "require_healthy"
                }
              ],
              "Connect": null,
              "Meta": null,
              "CanaryMeta": null,
              "Namespace": "default",
              "OnUpdate": "require_healthy",
              "Provider": "consul"
            }
          ],
          "Vault": null,
          "Templates": null,
          "Constraints": null,
          "Affinities": null,
          "Resources": {
            "CPU": 100,
            "Cores": 0,
            "MemoryMB": 50,
            "MemoryMaxMB": 0,
            "DiskMB": 0,
            "IOPS": 0,
            "Networks": null,
            "Devices": null
          },
          "RestartPolicy": {
            "Attempts": 2,
            "Interval": 1800000000000,
            "Delay": 15000000000,
            "Mode": "fail"
          },
          "DispatchPayload": null,
          "Lifecycle": null,
          "Meta": null,
          "KillTimeout": 5000000000,
          "LogConfig": {
            "MaxFiles": 10,
            "MaxFileSizeMB": 10
          },
          "Artifacts": null,
          "Leader": false,
          "ShutdownDelay": 0,
          "VolumeMounts": null,
          "ScalingPolicies": null,
          "KillSignal": "",
          "Kind": "",
          "CSIPluginConfig": null
        }
      ],
      "EphemeralDisk": {
        "Sticky": false,
        "SizeMB": 300,
        "Migrate": false
      },
      "Meta": null,
      "ReschedulePolicy": {
        "Attempts": 0,
        "Interval": 0,
        "Delay": 30000000000,
        "DelayFunction": "exponential",
        "MaxDelay": 3600000000000,
        "Unlimited": true
      },
      "Affinities": null,
      "Spreads": null,
      "Networks": [
        {
          "Mode": "",
          "Device": "",
          "CIDR": "",
          "IP": "",
          "Hostname": "",
          "MBits": 0,
          "DNS": null,
          "ReservedPorts": null,
          "DynamicPorts": [
            {
              "Label": "db",
              "Value": 0,
              "To": 6379,
              "HostNetwork": "default"
            }
          ]
        }
      ],
      "Consul": {
        "Namespace": ""
      },
      "Services": null,
      "Volumes": null,
      "ShutdownDelay": null,
      "StopAfterClientDisconnect": null,
      "MaxClientDisconnect": null
    }
  ],
  "Update": {
    "Stagger": 30000000000,
    "MaxParallel": 1,
    "HealthCheck": "",
    "MinHealthyTime": 0,
    "HealthyDeadline": 0,
    "ProgressDeadline": 0,
    "AutoRevert": false,
    "AutoPromote": false,
    "Canary": 0
  },
  "Multiregion": null,
  "Periodic": null,
  "ParameterizedJob": null,
  "Dispatched": false,
  "DispatchIdempotencyToken": "",
  "Payload": null,
  "Meta": null,
  "ConsulToken": "",
  "ConsulNamespace": "",
  "VaultToken": "",
  "VaultNamespace": "",
  "NomadTokenID": "",
  "Status": "running",
  "StatusDescription": "",
  "Stable": true,
  "Version": 13,
  "SubmitTime": 1667924294904831500,
  "CreateIndex": 22065,
  "ModifyIndex": 70933,
  "JobModifyIndex": 70933
}

**Nomad Autoscaler log:**
2022-11-09T17:09:23.649Z [ERROR] policy_manager.policy_handler: encountered an error monitoring policy: policy_id=0c66157a-0e7b-fdc9-de1d-5fb429d14439 error="failed to get policy: Unexpected response code: 404 (policy not found)"
2022-11-09T17:09:23.650Z [ERROR] internal_plugin.app-sizing-nomad: failed to call Nomad API: job_id=example namespace=default error="Unexpected response code: 404 (job not found)"
2022-11-09T17:09:23.650Z [TRACE] internal_plugin.app-sizing-nomad: stopping internal cache updater of task statuses: job_id=example namespace=default
2022-11-09T17:09:23.651Z [TRACE] policy_manager: received policy IDs listing: num=9 policy_source=nomad
2022-11-09T17:09:23.651Z [TRACE] policy_manager: handler already exists: policy_id=3d406e1e-e590-936b-a68a-a2d563f029ba policy_source=nomad
2022-11-09T17:09:23.651Z [TRACE] policy_manager: handler already exists: policy_id=0ad8e3f2-1554-608f-0128-b6b834e935aa policy_source=nomad
2022-11-09T17:09:23.651Z [TRACE] policy_manager: handler already exists: policy_id=3e211765-4432-a648-53e5-e1d89f16fba9 policy_source=nomad
2022-11-09T17:09:23.651Z [TRACE] policy_manager: handler already exists: policy_id=45c5dbce-30fd-400c-b98e-368ee2a403c7 policy_source=nomad
2022-11-09T17:09:23.651Z [TRACE] policy_manager: handler already exists: policy_id=22f6ab45-d16e-a7ff-838b-ac7b1e42ea0e policy_source=nomad
2022-11-09T17:09:23.651Z [TRACE] policy_manager: handler already exists: policy_id=33269fda-7bc6-5c0b-d312-86ff4e013b2b policy_source=nomad
2022-11-09T17:09:23.651Z [TRACE] policy_manager: handler already exists: policy_id=7a490cc1-c319-965d-3fed-10e226ba1341 policy_source=nomad
2022-11-09T17:09:23.651Z [TRACE] policy_manager: handler already exists: policy_id=a0b53cf7-069a-287e-76c1-6225bce15ea6 policy_source=nomad
2022-11-09T17:09:23.651Z [TRACE] policy_manager: handler already exists: policy_id=a9c83ddf-f0d0-e1c2-12f5-bcb1a61cb518 policy_source=nomad
2022-11-09T17:09:23.651Z [TRACE] policy_manager.policy_handler: stopping handler: policy_id=0c66157a-0e7b-fdc9-de1d-5fb429d14439
2022-11-09T17:09:23.651Z [TRACE] policy_manager.policy_handler: stopping policy handler due to done channel: policy_id=0c66157a-0e7b-fdc9-de1d-5fb429d14439
2022-11-09T17:09:23.651Z [TRACE] nomad_policy_source: done with policy monitoring: policy_id=0c66157a-0e7b-fdc9-de1d-5fb429d14439

**Comment by mikenomitch** [comment]
I think this might be the same as https://github.com/hashicorp/nomad-autoscaler/issues/519, but I'm not sure, so keeping both open for now.

**Comment by gulducat** [comment]
I was able to reproduce this pretty readily (yay!) by running and stop/purge-ing a job with a scaling policy.

This one is similar to 519 in that they're both related to blocking API calls to Nomad, but 519 is about what happens when the server that the autoscaler is connected to goes away, whereas this one is a result of a query actually behaving normally, just with an extra error log (presumably one that trips monitoring, rightly concerning operators).

Here the `/v1/scaling/policy/{specific-policy-id}` blocking query gets a 404 when a policy is deleted, so no longer exists, but that's actually an expected, reasonable event in the system, so should not be an error.

It's basically the result of a race condition between

1. The autoscaler policy Manager "monitor"s all the policies, and updates which individual ones to watch when the list changes
2. An individual policy watcher (called a [Handler](https://github.com/hashicorp/nomad-autoscaler/blob/v0.3.7/policy/handler.go#L26)) gets a 404 [here](https://github.com/hashicorp/nomad-autoscaler/blob/v0.3.7/policy/handler.go#L146), emitting the error

If `#1` won the race, we wouldn't get the error because the Handler would be stopped, but it's not trivial to coordinate multiple blocking queries like that (to force `#1` to always win), so I'm thinking of making `#2` just log an info event for 404s (for future possible debug purposes), then `#1` will kick in afterwards, since the list will also have changed, and clean up the Handler.

**Comment by gulducat** [comment]
This log line:

> [DEBUG] internal_plugin.nomad-target: failed to read job scale status, retrying: Job=example error="Unexpected response code: 404 (job not found)"

was a little trickier to catch, because it may appear up to 5 minutes after the job is purged, due to [this blocking query option](https://github.com/hashicorp/nomad-autoscaler/blob/v0.3.7/plugins/builtin/target/nomad/plugin/state.go#L160) combined with [this part of the Nomad API](https://github.com/hashicorp/nomad/blob/v1.6.2/nomad/job_endpoint.go#L2213) not considering it an error for the job to go away, so it doesn't stop the blocking query.

I'm inclined to keep the debug line there, as it is potentially useful (and not labeled an error).  I might update the Nomad side though, to unblock the query if the job goes away.

---

I have not seen this one in my testing:

> [ERROR] internal_plugin.app-sizing-nomad: failed to call Nomad API: job_id=example namespace=default error="Unexpected response code: 404 (job not found)"

I wonder if it might have been corrected by some other change before I started looking here, but have not yet confirmed that.
